UXENG-3184: Add bitbucket and gitlab scm docs to scm page (#454)
algorithmiaio/dev-center: _pages/algorithm-development/source-control-management.md

__I revoked access to the Algorithmia's Github OAuth application, and now none of the repositories I created can build!__

When you revoke access to an OAuth app, Github automatically revokes both your token and any deploy keys you may have created via the OAuth app. To fix your algorithms, simply [follow the instructions above](#your-repositorys-deploy-key-was-removed) on restoring deploy keys to a repository.

### Hosting Source Code on Bitbucket

By hosting your algorithm's source code on Bitbucket, you can take advantage of Bitbucket's rich set of developer features, such as pull requests, and also ensure that access to your source code is carefully mediated.

If you're new to Git, we recommend [this tutorial series](https://try.github.io/) by GitHub.

**Web IDE Support** At this time we do not support editing source code in our web app for Bitbucket-hosted algorithms. 
{: .notice-info}

#### Connecting Your Algorithmia & Bitbucket Accounts

To connect your Algorithmia and Bitbucket accounts, simply select the Bitbucket instance you wish to use when creating your algorithm:

> <img src="{{site.cdnurl}}{{site.baseurl}}/images/post_images/source_code_management/scm_create_bitbucket_algorithm_unauthorized.png" alt="Authorization prompt for GitHub repository host" class="screenshot img-sm">

Reviewer: ```suggestion
<img src="{{site.cdnurl}}{{site.baseurl}}/images/post_images/source_code_management/scm_create_bitbucket_algorithm_unauthorized.png" alt="Authorization prompt for Bitbucket repository host" class="screenshot img-sm">
```

By default, all Bitbucket respositories are created under your personal Bitbucket account. However, you may optionally choose any Bitbucket organizations that you are a member of. These organizations are listed in the "Repository Owner" dropdown:

<img src="{{site.cdnurl}}{{site.baseurl}}/images/post_images/source_code_management/scm_create_bitbucket_algorithm_org.png" alt="Customizing your repository owner" class="screenshot img-sm">

You may also customize the name of the repository that is created. By default the name of your algorithm is used. 

<img src="{{site.cdnurl}}{{site.baseurl}}/images/post_images/source_code_management/scm_create_bitbucket_algorithm_name.png" alt="Customizing your repository name" class="screenshot img-sm">

If you decide to use an alternative name, we recommend that you use only letters, numbers, hyphens, and/or underscores. While Bitbucket will accept characters outside this range, they will replace any unsupported characters with an underscore, resulting in an undesirable repository name.

When we create your repository, we associate the following with it:

> - **An Access Key**: Bitbucket deploy keys allow read-only access to specific repositories, and are GitHub’s prescribed means by which external services can fetch code for building and deploying. These keys are not tied to individual permissions, and as such will allow Algorithmia to continue building an algorithm even if the permissions of the creating user change. We also use deploy keys to obtain your repository's commit log (to display changes when publishing versions) and README.md (for use as algorithm documentation).

Reviewer: ```suggestion
- **An Access Key**: Bitbucket access keys allow read-only access to specific repositories, and are Bitbucket’s prescribed means by which external services can fetch code for building and deploying. These keys are not tied to individual permissions, and as such will allow Algorithmia to continue building an algorithm even if the permissions of the creating user change. We also use access keys to obtain your repository's commit log (to display changes when publishing versions) and README.md (for use as algorithm documentation).
```

{: .notice-info}

#### Troubleshooting Bitbucket-Hosted Algorithms

With the source for your algorithm hosted externally, there is a chance that our connection to Bitbucket could become disrupted. When this happens, it could be due to one of the following:

##### Your Repository's Access Key Was Removed

We depend on access keys to pull source code from your algorithm for building. A access key is simply an SSH keypair with read-only access to a specific repository. Algorithmia securely stores the private key while the public key is shared with Bitbucket.

You can obtain the public key for your repository by navigating to your algorithm's settings page and clicking "View Key":

You will then be presented with a modal, from which you can copy the public key.

With the access key in hand, simply follow [instructions outlined here](https://support.atlassian.com/bitbucket-cloud/docs/add-access-keys/) to restore the public key to your repository.

##### Your Repository Was Deleted

> If your algorithm's Bitbucket repository was deleted, Bitbucket does not have a mechanism for restoring. It is essential that both the repository's deploy key and webhooks remain in place, otherwise we will not be able to pull your source code, or know when you make changes.

Reviewer: ```suggestion
If your algorithm's Bitbucket repository was deleted, Bitbucket does not have a mechanism for restoring. It is essential that both the repository's access key and webhooks remain in place, otherwise we will not be able to pull your source code, or know when you make changes.
```

##### Your Repository's Webhook Needs to be Restored

> We depend on webhooks to notify us when changes have occurred to your algorithm's repository. To restore your webhook within Bitbucket, first locate the webhook URL and secret. This can be found on your algorithm's settings page by clicking 'View Details' under 'Bitbucket Webhook'.

Reviewer: ```suggestion
We depend on webhooks to notify us when changes have occurred to your algorithm's repository. To restore your webhook within Bitbucket, first locate the webhook URL. This can be found on your algorithm's settings page by clicking 'View Details' under 'Bitbucket Webhook'.
```

> You will then be presented with a modal, from which you can copy the webhook URL and secret.

Reviewer: ```suggestion
You will then be presented with a modal, from which you can copy the webhook URL.
```

> Once you've found these values, visit your algorithm's repository on Bitbucket and click on the Settings tab. Click 'Webhooks' on the left side, and you'll be able to restore your webhook by entering the webhook URL and secret.

Reviewer: The endpoints for loading SCM data weren't returning secrets when I was testing Bitbucket endpoints. I'd leave that part off for now. Looking in BitBucket's UI I didn't see secrets mentioned anywhere.
```suggestion
Once you've found this value, visit your algorithm's repository on Bitbucket and click on the Settings tab. Click 'Webhooks' on the left side, and you'll be able to restore your webhook by entering the webhook URL.
```

#### FAQ

__I'd like to reuse an existing Bitbucket repository for my algorithm. Is this possible?__

At this time we are only able to use new, purpose-created repositories for Bitbucket-backed algorithms, which Algorithmia itself creates and provisions.

__Can I rename or move the Bitbucket repository that backs my algorithm?__

Yes! Upon any change to your repository's name or owner we receive a webhook which we will use to update our system to point to the new repository address.

__I revoked access to the Algorithmia's Bitbucket OAuth application, and now none of the repositories I created can build!__

> When you revoke access to an OAuth app, Bitbucket automatically revokes both your token and any deploy keys you may have created via the OAuth app. To fix your algorithms, simply [follow the instructions above](#your-repositorys-access-key-was-removed) on restoring deploy keys to a repository.

Reviewer: ```suggestion
When you revoke access to an OAuth app, Bitbucket automatically revokes both your token and any access keys you may have created via the OAuth app. To fix your algorithms, simply [follow the instructions above](#your-repositorys-access-key-was-removed) on restoring deploy keys to a repository.
```

By default, all GitLab respositories are created under your personal GitLab account. However, you may optionally choose any GitLab organizations that you are a member of. These organizations are listed in the "Repository Owner" dropdown:

<img src="{{site.cdnurl}}{{site.baseurl}}/images/post_images/source_code_management/scm_create_gitlab_algorithm_org.png" alt="Customizing your repository owner" class="screenshot img-sm">

You may also customize the name of the repository that is created. By default the name of your algorithm is used. 

<img src="{{site.cdnurl}}{{site.baseurl}}/images/post_images/source_code_management/scm_create_gitlab_algorithm_name.png" alt="Customizing your repository name" class="screenshot img-sm">

If you decide to use an alternative name, we recommend that you use only letters, numbers, hyphens, and/or underscores. While GitLab will accept characters outside this range, they will replace any unsupported characters with an underscore, resulting in an undesirable repository name.

When we create your repository, we associate the following with it:

> - **A Deploy Key**: GitLab deploy keys allow read-only access to specific repositories, and are GitHub’s prescribed means by which external services can fetch code for building and deploying. These keys are not tied to individual permissions, and as such will allow Algorithmia to continue building an algorithm even if the permissions of the creating user change. We also use deploy keys to obtain your repository's commit log (to display changes when publishing versions) and README.md (for use as algorithm documentation).

Reviewer: ```suggestion
- **A Deploy Key**: GitLab deploy keys allow read-only access to specific repositories, and are GitLab’s prescribed means by which external services can fetch code for building and deploying. These keys are not tied to individual permissions, and as such will allow Algorithmia to continue building an algorithm even if the permissions of the creating user change. We also use deploy keys to obtain your repository's commit log (to display changes when publishing versions) and README.md (for use as algorithm documentation).
```

- **Webhooks**: We set up webhooks to receive notifications about changes to your repo, such as when there's a change to its default branch.

#### Updating GitLab-Hosted Algorithms

Once you've created your algorithm, any commits to your GitLab repository's _default_ branch will result in a build on Algorithmia. At creation, your repository's default branch will be your `master` branch, but you can change this at any time in your repository's settings.

> You can view your algorithm's builds by heading to it's landing page and clicking the "Builds" tab if you are the algorithm's owner.

Reviewer: ```suggestion
You can view your algorithm's builds by heading to its landing page and clicking the "Builds" tab if you are the algorithm's owner.
```
